text/tls-and-alpn.rst

```diff
@@ -0,0 +1,380 @@
+   solve the remote client trouble though.
+2. Provide a client-side switch that the user could easily set, so that the
+   user could explicitly turn it on for development mode, and off for
+   production.
+
+
+Production Scenario
+^^^^^^^^^^^^^^^^^^^
+
+Question: do we want to manage the certificate for the user? That means to
+import the user certificate files and store locally, warn about expiry or even
+run certbot to renew some free certs (this might be too much). If yes, the
+following in this section is just one of the options.
+
+``edgedb server init`` will have additional parameters to import the TLS
+certificate and key files for the new EdgeDB instance::
+
+    --import-tls-certfile
+    --import-tls-keyfile
+    --tls-private-key-password
+
+Again the password option here yields a security concern as mentioned
+previously. We should probably follow the way PostgreSQL used to make it a
+customizable command to obtain the passphrase.
+
+For simplicity, optional TLS settings are not provided on ``server init``.
+These options will also be added to ``edgedb project init`` command.
+
+In addition, we should provide a new ``edgedb server update-tls`` command for
+the user to renew the certificate or tweak optional TLS settings. For example::
+
+    --import-tls-certfile
+    --import-tls-keyfile
+    --set-private-key-password
+    --set-protocol-min
+    --set-protocol-max
+    --set-ciphers
+    --generate-dev-cert
+
+
+ALPN and Protocol Changes
+-------------------------
+
+This chapter is about the implementation detail of how to use ALPN and what is
+changed in the protocol. I have 2 options:
+
+
+Option 1: Only ``binary`` and ``http``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The server advertises only two ALPN protocols: ``edgedb-binary`` and ``http``,
+where ``edgedb-binary`` is the EdgeDB binary protocol, and ``http`` includes
+all the other HTTP-based protocols. As the current code base is already serving
+different protocols like EdgeQL and GraphQL over HTTP by different URL path
+prefix on the same HTTP port, Option 1 here has the minimum effort to change.
+
+The client should choose between ``edgedb-binary`` and ``http`` based on the
+scenario the user is using the client. If the client didn't join the protocol
+negotiation, the server will fallback to ``http`` - it is literally just HTTPS.
+
+There will be no protocol change, as we're only adding an mandatory TLS, and
+replacing the magic first byte check with ALPN.
+
+
+Option 2: Fine-grained Protocols
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+All protocols/extensions are given a unique ALPN protocol name, so that the URI
+can be largely simplified. Other than that, the actual protocols remain
+unchanged. The server will advertise all ALPN protocols for simplicity, even
+though some extension may never be configured on any database.
+
+If for some reason the ALPN negotiation didn't work, the server will fallback
```

Comment:
What can be a reason for ALPN to not work?

Reply:
Oh sorry - this is an example where a 3rd-party client didn't specify any ALPN protocol and the negoation was a failure.